etest_test/build_test.py

```diff
@@ -9,4 +14,193 @@ def test_help() -> None:
+    container_remove = mocker.patch("etest.build.docker.container.remove")
+
+    # Run the program
+    runner = click.testing.CliRunner()
+    result = runner.invoke(sut.main, ["--libc", libc, hardened, "--verbosity", "DEBUG"])
+
+    # Check for errors
     assert result.exit_code == 0
+
+    # Assert all functions have been called
+    build.assert_called_once()
+    run.assert_called_once()
+    commit.assert_called_once()
+    image_remove.assert_called_once()
+    container_remove.assert_called_once()
+
+
+def test_builderror(mocker: MockerFixture) -> None:
+    """Ensure the program fails on a BuildError."""
+    # Raise a ContainerError
+    log = itertools.tee({"error": "mockup"})[0]
+    error = docker.errors.BuildError(reason="Mockup error.", build_log=log)
+
+    build = mocker.patch(
+        "etest.build.docker.image.build",
+        side_effect=error,
+    )
+
+    # Run the program
+    runner = click.testing.CliRunner()
+    result = runner.invoke(sut.main, ["--verbosity", "DEBUG"])
+
+    # Check if it triggered an error
+    assert result.exit_code == 1
+
+    # Assert all functions have been called
+    build.assert_called_once()
+
+
+def test_cleanup(mocker: MockerFixture) -> None:
+    """Ensure cleaning happens after exceptions."""
+    # Mock build functions
+    build = mocker.patch(
+        "etest.build.docker.image.build",
+        return_value=True,
+    )
+
+    # Raise a ContainerError
+    error = docker.errors.ContainerError(
+        container=True,
+        exit_status=1,
+        command="ls",
+        image="ebuildtest/etest",
+        stderr=None,
+    )
+    run = mocker.patch(
+        "etest.build.docker.container.run",
+        side_effect=error,
+    )
+
+    # Mock cleanup functions
+    image_remove = mocker.patch("etest.build.docker.image.remove")
+    container_remove = mocker.patch("etest.build.docker.container.remove")
+
+    # Run the program
+    runner = click.testing.CliRunner()
+    result = runner.invoke(sut.main, ["--verbosity", "DEBUG"])
+
+    # Check if it triggered an error
+    assert result.exit_code == 1
+
+    # Assert all functions have been called
+    build.assert_called_once()
+    run.assert_called_once()
+    image_remove.assert_called_once()
+    container_remove.assert_called_once()
+
+
+def test_push(mocker: MockerFixture) -> None:
+    """Ensure etest-build can push."""
+    # Mock push functions
+    push = mocker.patch(
+        "etest.build.docker.image.push",
+        return_value="pushing...\nfinished.",
+    )
+
+    # Run the program
+    runner = click.testing.CliRunner()
+    result = runner.invoke(sut.main, ["--no-build", "--push", "--verbosity", "DEBUG"])
+
+    # Check for errors
+    assert result.exit_code == 0
+
+    # Check if logging is handled correctly
+    out_debug = [o for o in result.output.splitlines() if "debug" in o]
+
+    assert "debug: pushing..." in out_debug
+    assert "debug: finished." in out_debug
```

Comment:
Could we more directly use endswith on the original output for a multiline string?  That's direct and indicates the expected behaviour, right?

Reply:
Although based on the name this is the wrong thing to assert.  Instead you simply should assert push.assert_called_once.

Reply:
Should I check for this in a different test instead?

Reply:
I'd argue we shouldn't ever test that our debug logging does anything.  It's debug (I read it as throw away information).  If we're asserting debug information we're not checking the actual behaviour of the system but a side effect of it that may or may not be correct.